src/main/java/org/logstash/beats/ThunderingGuardHandler.java

```diff
@@ -0,0 +1,40 @@
+package org.logstash.beats;
+
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This handler is responsible to avoid accepting new connections when the direct memory
+ * consumption is close to the MaxDirectMemorySize.
+ * <p>
+ * If the total allocated direct memory is close to the max memory size and also the pinned
+ * bytes from the direct memory allocator is close to the direct memory used, then it drops the new
+ * incoming connections.
+ * */
+@Sharable
+public final class ThunderingGuardHandler extends ChannelInboundHandlerAdapter {
+
+    private final static long MAX_DIRECT_MEMORY = io.netty.util.internal.PlatformDependent.maxDirectMemory();
+
+    private final static Logger logger = LogManager.getLogger(ThunderingGuardHandler.class);
+
+    @Override
+    public void channelRegistered(final ChannelHandlerContext ctx) throws Exception {
+        PooledByteBufAllocator pooledAllocator = (PooledByteBufAllocator) ctx.alloc();
+        long usedDirectMemory = pooledAllocator.metric().usedDirectMemory();
+        if (usedDirectMemory > MAX_DIRECT_MEMORY * 0.90) {
+            long pinnedDirectMemory = pooledAllocator.pinnedDirectMemory();
+            if (pinnedDirectMemory >= usedDirectMemory * 0.80) {
+                ctx.close();
+                logger.info("Dropping connection {} due to high resource consumption", ctx.channel());
```

Comment:
should this be the order? also it's likely worth bumping to warn as it's a near OOM situation.

Reply:
It's almost idempotent, but in e general we could have that pinned is close to used, but far from max direct. In such a case I think don't want to enter too much frequently the check on max memory.
For example considering:
- max direct is 128MB
- used is 64MB
- pinned is constantly at 56 MB
The system is balanced for the work it's doing to consume ~50MB and it's far from OOM condition, so it's not in harmful situation, I think.